design/jwt-verification-design.md

```diff
@@ -0,0 +1,471 @@
+# JWT Verification Support
+
+## Abstract
+This document describes a design for performing JSON Web Token (JWT) verification for requests to virtual hosts hosted by Contour.
+
+## Background
+JSON Web Token (JWT) is an open standard (RFC 7519) that defines a compact and self-contained way for securely transmitting information between parties as a JSON object. 
+This information can be verified and trusted because it is digitally signed. (ref. https://jwt.io/introduction)
+
+Envoy Proxy has built-in support for verifying JWTs that are attached to incoming requests, via the [JWT Authentication HTTP filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/jwt_authn_filter#config-http-filters-jwt-authn). (Note, this document will use the term **JWT verification** to describe this process).
+Specifically, Envoy can verify the signature, audience, issuer and time restrictions of a JWT.
+If verification fails, the request will be rejected.
+It's important to note that this filter does not itself obtain a JWT for an incoming request; the request must already have one attached.
+
+Contour does not currently have support for configuring the JWT authentication filter.
+This document proposes a design for adding that support to Contour's custom resource, `HTTPProxy`.
+
+## Goals
+- JWT verification for requests to TLS-enabled virtual hosts.
+- Expose a subset of the Envoy filter configuration to cover the most common use cases.
+- Be able to easily expose additional configuration settings if/when needed in the future.
+
+## Non Goals
+- JWT verification for requests to non-TLS enabled virtual hosts.
+- Exposing all possible Envoy configuration settings.
+- Supporting end-to-end OAuth2/OIDC flows.
+
+## High-Level Design
+Contour's `HTTPProxy` resource will get a new optional field, `spec.virtualhost.jwtProviders`, to define the details of how to verify JWTs for requests to a given virtual host.
+This field will only be supported for virtual hosts for which Envoy is terminating TLS.
+The structure of this field will be similar to the [Envoy filter's providers field](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#envoy-v3-api-msg-extensions-filters-http-jwt-authn-v3-jwtauthentication), with some simplifications. 
+Specifically, a provider will define an issuer, 0+ audiences, and a JSON Web Key Set (JWKS) that can be used to verify a JWT (see [the Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#envoy-v3-api-msg-extensions-filters-http-jwt-authn-v3-jwtprovider) for more information).
+Any number of providers can be defined for a virtual host, to allow different routes to be verified differently.
+At most one provider can be marked as the "default", meaning it will automatically be applied as a requirement to all routes unless they explicitly opt out (more below).
```

Comment:
What do you think if instead of separate field `default: true` we would pick the default provider by name `name: default`. It would avoid the error case where more than one providers have `default: true` set. However, the provider named `default` then becomes "magically" applied by default, which of course could be considered as bit surprising by users? I'm unsure myself. Maybe putting a special meaning to a name is a bad idea, but I wanted to hear what you think?

Reply:
Yeah, I've definitely used that pattern before for similar situations -- it does have the advantage of not requiring a separate flag which is nice. When I used it before, it was for the names of CR's themselves, which meant that the API server would prevent multiple resources from being created with the name "default"; in this case, though, a user could name multiple JWT providers "default" so we'd still need code to check for that situation & set an error condition. I think since we'd have to do that runtime check regardless, I probably lean towards keeping the flag as I think it is a little less "magical", but open to others' thoughts here too.